NixOS/nixpkgs
# GPU access in the sandbox

## Description of changes

The PR introduces a way to expose devices and drivers (e.g. GPUs and `/run/opengl-driver/lib`) in the Nix build's sandbox for specially marked derivations. The approach is that described in https://github.com/NixOS/nixpkgs/issues/225912: we introduce a pre-build-hook which would scan the `.drv` being built for the presence of a chosen tag in the `requiredSystemFeatures` attribute and, upon a match, instruct Nix to mount a list of device-related paths. E.g. one could mark the derivation as `mkDerivation { ...; requiredSystemFeatures = [ "cuda" ]; }` to indicate that it would need to use CUDA during the build. Untagged derivations would still observe "pure" environment without any devices. Further, the `requiredSystemFeatures` attribute is special, Nix uses it to determine whether a host is capable of building the derivation, and which remote builders to choose. A host that hasn't been set up with `nix.settings.system-features` (`system-features` in `nix.conf(5)`) would refuse to build the marked derivation

One application of this hook is to implement GPU tests as normal derivations (cf. the torch example in the diff).
Another use could be to use Nix to run ML/DL/NumberCrunching pipelines that utilize GPUs (e.g. https://github.com/SomeoneSerge/pkgs/blob/5bf5ea2b16696e70ade5a68063042dcf2e8f033b/python-packages/by-name/om/omnimotion/package.nix#L270-L291). Similarly, one could write derivations that render things using OpenGL or Vulkan

How to test:

1. Take a NixOS machine with an nvidia GPU and `hardware.opengl.enable = true`.
2. Try building `python3Packages.torch.tests.cudaAvailable`, observe it's refused evaluation
3. Rebulid with `programs.nix-required-mounts.enable = true` (

    <details>

    ```
    # Note: nix-required-sandbox-paths extends system-features,
    # but they won't be merged with the nixos defaults, because they got a different priority
    # -> set manually:
    nix.settings.system-features = [
      "nixos-test"
      "benchmark"
      "big-parallel"
      "kvm"
    ];
    programs.nix-required-sandbox-paths.enable = true;
    ```
   
    </details>
4. Build `python3Packages.torch.gpuChecks.cudaAvailable` successfully
5. Try dropping `cuda` from `gpuChecks.cudaAvailable`'s `requiredSystemFeatures`, and observe the test fail

CC @NixOS/cuda-maintainers 



## Alternatives and related work

@tomberek suggests that one could have derivations ask for specific paths using `__impureHostDeps`, currently undocumented and being used for Darwin builds

For GPU tests, @Kiskae suggests one could instead use NixOS tests with the PCI passthrough

We could also possibly consider supporting variables like `CUDA_VISIBLE_DEVICES` and/or re-using nvidia-docker/nvidia-container-toolkit because that's the interface many (singularity, docker) rely on. These currently do not support any static configuration, and in particular they rely on glibc internals to discover the host drivers in a very non-cross-platform way.

## Things done



- Built on platform(s)
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
